components/playlist/playlist_media_file_downloader.h

```diff
@@ -0,0 +1,127 @@
+                                  const std::string& media_file_path_key,
+                                  const std::string& media_file_path) = 0;
+    // Called when target media file generation failed.
+    virtual void OnMediaFileGenerationFailed(const std::string& id) = 0;
+
+   protected:
+    virtual ~Delegate() {}
+  };
+
+  PlaylistMediaFileDownloader(
+      Delegate* delegate,
+      content::BrowserContext* context,
+      base::FilePath::StringType source_media_files_dir,
+      base::FilePath::StringType unified_media_file_name,
+      std::string media_file_path_key,
+      std::string create_params_path_key);
+  virtual ~PlaylistMediaFileDownloader();
+
+  PlaylistMediaFileDownloader(const PlaylistMediaFileDownloader&) = delete;
+  PlaylistMediaFileDownloader& operator=(const PlaylistMediaFileDownloader&) =
+      delete;
+
+  void GenerateSingleMediaFile(base::Value&& playlist_value,
+                               const base::FilePath& base_dir);
+
+  void RequestCancelCurrentPlaylistGeneration();
+
+  bool in_progress() const { return in_progress_; }
+  std::string current_playlist_id() const { return current_playlist_id_; }
+
+ private:
+  using SimpleURLLoaderList =
+      std::list<std::unique_ptr<network::SimpleURLLoader>>;
+
+  void ResetDownloadStatus();
+  void DownloadAllMediaFileSources();
+  void DownloadMediaFile(const GURL& url, int index);
+  void CreateSourceFilesDirThenDownloads();
+  void OnSourceFilesDirCreated(bool success);
+  void OnMediaFileDownloaded(SimpleURLLoaderList::iterator iter,
+                             int index,
+                             base::FilePath path);
+  void StartSingleMediaFileGeneration();
+  void OnSingleMediaFileGenerated(const std::string& id, bool success);
+
+  // True when all source media files are downloaded.
+  // If it's true, single media file will be generated.
+  bool IsDownloadFinished();
+  int GetNumberOfMediaFileSources();
+
+  void NotifyFail(const std::string& id);
+  void NotifySucceed(const std::string& id,
+                     const std::string& media_file_path_key,
+                     const std::string& media_file_path);
+
+  base::SequencedTaskRunner* task_runner();
+
+  Delegate* delegate_;
+  scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory_;
+
+  const base::FilePath::StringType source_media_files_dir_;
+  const base::FilePath::StringType unified_media_file_name_;
+  const std::string media_file_path_key_;
+  const std::string create_params_path_key_;
+
+  // All below variables are only for playlist creation.
+  base::FilePath playlist_dir_path_;
+  base::Value current_playlist_;
+  std::string current_playlist_id_;
+  int remained_download_files_ = 0;
+  int media_file_source_files_count_ = 0;
+
+  // true when this class is working for playlist now.
+  bool in_progress_ = false;
+
+  scoped_refptr<base::SequencedTaskRunner> task_runner_;
+
+  SimpleURLLoaderList url_loaders_;
```

Comment:
Can we use `api_request_helper::APIRequestHelper` in this PR instead of using `SimpleURLLoader` directly?

Reply:
Cool! I made a more API to APIRequestHelper for it.

Reply:
Wow, cool! 👍🏼